### PR TITLE
Fix last DLT problem(s) with the new backend.

### PR DESCRIPTION
**Pull request by abergeron**
This fixes a number of problems.

The main one is that the lifter for Pool was removing operations from the graph in certain circumstances.

Others are:
 - The lifter for Print was mostly broken
 - AbstractConv could stay unoptimized in certain cases of mixed input types.

There is now code in the AbstractConv ops to convert all inputs to be of the same type as the first one, which avoids the second problem.

### CODE REVIEW

File: theano/sandbox/gpuarray/opt.py [hunk 1]
```diff
@@ -786,10 +783,9 @@ def local_gpua_softmaxwithbias(node, context_name):
 @register_opt('fast_compile')
 @op_lifter([theano.tensor.opt.Assert])
 def local_assert(node, context_name):
-    if (node.inputs[0].owner and
-            isinstance(node.inputs[0].owner.op, HostFromGpu)):
-        return [host_from_gpu(node.op(node.inputs[0].owner.inputs[0],
-                                      *node.inputs[1:]))]
+    return [host_from_gpu(node.op(as_gpuarray_variable(node.inputs[0],
+                                                       context_name),
+                                  node.inputs[1:]))]
```
Comment (lamblin) [inline theano/sandbox/gpuarray/opt.py]: Is the removal of the `*` here intended? I don't think it makes sense.

**Comment by nouiz** [comment]
travis have this error:

/home/travis/build/Theano/Theano/theano/tensor/nnet/abstract_conv.py:641:32: E128 continuation line under-indented for visual indent

**Comment by nouiz** [comment]
Ok to merge when the flake8 test is fixed and you have run locally the tests under sandbox/gpuarray

**Comment by abergeron** [comment]
Locally, the everfailing scan test fails, but everything else passes (Although I did not test with the flake8 correction).

**Comment by nouiz** [comment]
we can merge when travis pass.

**Pull request merged**